File: android_inference_benchmark/app/src/main/java/com/visionforge/mobile/domain/video/DecodedFrameMetadataLedger.java
package com.visionforge.mobile.domain.video;

import java.util.Optional;

/** APP 端解码后新鲜度账本；规则与 Host C++ DecodedFrameMetadataLedger 对称。 */
public final class DecodedFrameMetadataLedger {
    public enum Disposition {
        FRESH,
        EPOCH_ADVANCED,
        DUPLICATE,
        SEQUENCE_CONFLICT,
        STALE_SEQUENCE,
        STALE_EPOCH,
        NOT_FRESH
    }

    public record Metadata(
            long streamEpoch,
            long frameSequence,
            long contentFingerprint,
            boolean idr,
            boolean repeat,
            boolean contentUpdated,
            long decodedAtNanos) {
        public Metadata {
            if (streamEpoch <= 0L || frameSequence < 0L
                    || frameSequence > VideoFragmentHeader.UINT32_MAX || decodedAtNanos < 0L) {
                throw new IllegalArgumentException("invalid decoded-frame metadata");
            }
        }
    }

    public record Observation(Disposition disposition, boolean acceptedForInference,
                              boolean epochChanged) {}

    private long epoch;
    private Long lastObservedSequence;
    private Metadata lastObserved;
    private Metadata lastFresh;
    private long transitionId;

    public synchronized Observation observe(Metadata metadata) {
        if (epoch != 0L && metadata.streamEpoch() < epoch) {
            return new Observation(Disposition.STALE_EPOCH, false, false);
        }
        boolean changed = false;
        if (epoch == 0L || metadata.streamEpoch() > epoch) {
            epoch = metadata.streamEpoch();
            lastObservedSequence = null;
            lastObserved = null;
            advanceTransition();
            lastFresh = null;
            changed = true;
        }
        if (lastObservedSequence != null) {
            if (metadata.frameSequence() < lastObservedSequence.longValue()) {
                return new Observation(Disposition.STALE_SEQUENCE, false, changed);
            }
            if (metadata.frameSequence() == lastObservedSequence.longValue()) {
                if (lastObserved != null
                        && (metadata.contentFingerprint() != lastObserved.contentFingerprint()
                        || metadata.idr() != lastObserved.idr()
                        || metadata.repeat() != lastObserved.repeat()
                        || metadata.contentUpdated() != lastObserved.contentUpdated())) {
                    // 同一身份的语义必须逐字段一致；否则 Host/APP 可能分别把它解释为
                    // repeat、预测帧或 IDR，属于必须 fail-closed 的跨语言协议冲突。
                    return new Observation(Disposition.SEQUENCE_CONFLICT, false, changed);
                }
                return new Observation(Disposition.DUPLICATE, false, changed);
            }
        }
        lastObservedSequence = metadata.frameSequence();
        lastObserved = metadata;
        if (metadata.repeat() || !metadata.contentUpdated()) {
            return new Observation(Disposition.NOT_FRESH, false, changed);
        }
        lastFresh = metadata;
        return new Observation(changed ? Disposition.EPOCH_ADVANCED : Disposition.FRESH, true, changed);
    }

    public synchronized Optional<Metadata> lastFresh() {
        return Optional.ofNullable(lastFresh);
    }

    public synchronized long transitionId() {
        return transitionId;
    }

    private void advanceTransition() {
        if (transitionId == Long.MAX_VALUE) {
            throw new IllegalStateException("decoded-frame transition id space exhausted");
        }
        transitionId++;
    }
}

File: dual_machine_runtime/e2e_core/src/host/domain/decoded_frame_metadata_ledger.cpp
#include "vf/host/domain/decoded_frame_metadata_ledger.hpp"

#include <limits>

namespace vf::host::domain {

FrameLedgerResult DecodedFrameMetadataLedger::observe(
    const DecodedFrameMetadata& metadata) noexcept {
    if (!metadata.identity.valid()) {
        return {.disposition = FrameDisposition::Invalid};
    }

    if (current_epoch_ && metadata.identity.stream_epoch < *current_epoch_) {
        return {.disposition = FrameDisposition::RetiredEpoch};
    }

    const bool new_epoch = !current_epoch_ || metadata.identity.stream_epoch > *current_epoch_;
    if (new_epoch && (metadata.is_repeat || !metadata.content_updated || !metadata.is_idr)) {
        // 只有真正解码出的新鲜 IDR 能建立一个新 epoch。重复帧、未更新画面或 P/B 帧
        // 都不能凭更大的 epoch 清空当前高水位。
        return {.disposition = FrameDisposition::EpochRequiresIdr};
    }

    bool epoch_changed = false;
    if (new_epoch) {
        current_epoch_ = metadata.identity.stream_epoch;
        last_observed_sequence_.reset();
        last_observed_frame_.reset();
        last_fresh_frame_.reset();
        if (transition_id_ != std::numeric_limits<std::uint64_t>::max()) {
            ++transition_id_;
        }
        epoch_changed = true;
    }

    if (last_observed_sequence_) {
        if (metadata.identity.frame_sequence < *last_observed_sequence_) {
            return {.disposition = FrameDisposition::StaleSequence,
                    .epoch_changed = epoch_changed};
        }
        if (metadata.identity.frame_sequence == *last_observed_sequence_) {
            if (last_observed_frame_ &&
                (metadata.content_fingerprint != last_observed_frame_->content_fingerprint ||
                 metadata.is_idr != last_observed_frame_->is_idr ||
                 metadata.is_repeat != last_observed_frame_->is_repeat ||
                 metadata.content_updated != last_observed_frame_->content_updated)) {
                // 同一帧身份的语义字段也必须完全一致。否则一端可能把它当 repeat，
                // 另一端却当成新鲜 IDR，属于必须立即关闭链路的协议冲突。
                return {.disposition = FrameDisposition::SequenceConflict,
                        .epoch_changed = epoch_changed};
            }
            return {.disposition = FrameDisposition::Duplicate,
                    .epoch_changed = epoch_changed};
        }
    }

    last_observed_sequence_ = metadata.identity.frame_sequence;
    last_observed_frame_ = metadata;

    if (metadata.is_repeat || !metadata.content_updated) {
        return {.disposition = FrameDisposition::NotFresh,
                .accepted_for_inference = false,
                .epoch_changed = epoch_changed};
    }

    last_fresh_frame_ = metadata;
    return {.disposition = epoch_changed ? FrameDisposition::EpochAdvanced
                                         : FrameDisposition::Fresh,
            .accepted_for_inference = true,
            .epoch_changed = epoch_changed};
}

}  // namespace vf::host::domain
